Library/OcCpuLib/OcCpuLib.c

```diff
@@ -107,6 +88,161 @@ ScanThreadCount (
+      &Context
+      );
+
+    if (EFI_ERROR (Status)) {
+      //
+      // It might make sense to error and exit here.
+      //
+      continue;
+    }
+
+    if (Context.PackageNumber + 1 >= Cpu->PackageCount) {
+      Cpu->PackageCount = (UINT16) (Context.PackageNumber + 1);
+    }
+
+    //
+    // According to the FrameworkMpServices header, EFI_MP_PROC_CONTEXT.NumberOfCores is the
+    // zero-indexed physical core number for the current processor. However, Apple appears to
+    // set this to the total number of physical cores (observed on Xserve3,1 with 2x Xeon X5550).
+    //
+    // This number may not be accurate; on MacPro5,1 with 2x Xeon X5690, NumberOfCores is 16 when
+    // it should be 12 (even though NumberOfProcessors is correct). Regardless, CoreCount and
+    // ThreadCount will be corrected in ScanIntelProcessor.
+    //
+    // We will follow Apple's implementation, as the FrameworkMpServices fallback was added for
+    // legacy Macs.
+    //
+    if (Context.NumberOfCores >= Cpu->CoreCount) {
+      Cpu->CoreCount = (UINT16) (Context.NumberOfCores);
+    }
+
+    //
+    // Similarly, EFI_MP_PROC_CONTEXT.NumberOfThreads is supposed to be the zero-indexed logical
+    // thread number for the current processor. On Xserve3,1 this was set to 2 (presumably to
+    // indicate that there are 2 threads per physical core).
+    //
+    if (Context.NumberOfCores * Context.NumberOfThreads >= Cpu->ThreadCount) {
+      Cpu->ThreadCount = (UINT16) (Context.NumberOfCores * Context.NumberOfThreads);
+    }
+  }
+
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+ScanThreadCount (
+  OUT OC_CPU_INFO  *Cpu
+  )
+{
+  EFI_STATUS                          Status;
+  EFI_MP_SERVICES_PROTOCOL            *MpServices;
+  FRAMEWORK_EFI_MP_SERVICES_PROTOCOL  *FrameworkMpServices;
+  UINTN                               NumberOfProcessors;
+  UINTN                               NumberOfEnabledProcessors;
+
+  Cpu->PackageCount = 1;
+  Cpu->CoreCount    = 1;
+  Cpu->ThreadCount  = 1;
+  NumberOfProcessors = 0;
+  NumberOfEnabledProcessors = 0;
+
+  Status = gBS->LocateProtocol (
+    &gEfiMpServiceProtocolGuid,
+    NULL,
+    (VOID **) &MpServices
+    );
+
+  if (EFI_ERROR (Status)) {
+    Status = gBS->LocateProtocol (
+      &gFrameworkEfiMpServiceProtocolGuid,
+      NULL,
+      (VOID **) &FrameworkMpServices
+      );
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "OCCPU: No MP services - %r\n", Status));
+      return Status;
+    } else {
+      Status = ScanFrameworkMpServices (
+        FrameworkMpServices,
+        Cpu,
+        &NumberOfProcessors,
+        &NumberOfEnabledProcessors
+        );
+    }
+  } else {
+    Status = ScanMpServices (
+      MpServices,
+      Cpu,
+      &NumberOfProcessors,
+      &NumberOfEnabledProcessors
+      );
+  }
+
+  DEBUG ((
+    DEBUG_INFO,
+    "OCCPU: MP services threads %u (enabled %u) - %r\n",
+    (UINT32) NumberOfProcessors,
+    (UINT32) NumberOfEnabledProcessors,
```

Comment:
Can't we just print UINT64 here? I know we will never realistically hit it, however I prefer to not have needless truncations and punning (as this is DEBUG-only, no shipping code will be hurt peformance-wise).